### PR TITLE
New organization for URLConnectionTest test case

### PR DESCRIPTION
**Pull request by klebertarcisio**
Hi everyone,

I am proposing a new organization for the URLConnectionTest class. I created 7 new classes with the following test cases:

URLConnectionProxyTest
connectVia**Proxy**UsingProxyArg
connectVia**Proxy**UsingProxySystemProperty
connectVia**Proxy**UsingHttpProxySystemProperty
connectViaHttp**Proxy**ToHttpsUsingProxyArgWithNoProxy
connectViaHttp**Proxy**ToHttpsUsingHttpProxySystemProperty
connectViaHttp**Proxy**ToHttpsUsingProxyArg
connectViaHttp**Proxy**ToHttpsUsingProxySystemProperty
connectViaHttp**Proxy**ToHttpsUsingHttpsProxySystemProperty
redirectWith**Proxy**Selector
response305Use**Proxy**

### URLConnectionStreamTest
flushAfter**Stream**TransmittedWithChunkedEncoding
flushAfter**Stream**TransmittedWithFixedLength
flushAfter**Stream**TransmittedWithNoLengthHeaders
input**Stream**AvailableWithChunkedEncoding
input**Stream**AvailableWithContentLengthHeader
input**Stream**AvailableWithNoLengthHeaders

### URLConnectionUploadTest
chunked**Upload**_byteByByte
chunked**Upload**_smallBuffers
chunked**Upload**_largeBuffers
fixedLength**Upload**_byteByByte
fixedLength**Upload**_smallBuffers
fixedLength**Upload**_largeBuffers

### URLGzipConnectionTest
**gzip**AndConnectionReuseWithFixedLength
**gzip**AndConnectionReuseWithChunkedEncoding
**gzip**AndConnectionReuseWithFixedLengthAndTls
**gzip**AndConnectionReuseWithChunkedEncodingAndTls

### URLResponseConnectionTest
**response**307WithGet
**response**307WithHead
**response**307WithOptions
**response**307WithPost
**response**308WithGet
**response**308WithHead
**response**308WithOptions
**response**308WithPost
**response**307WithPostReverted
**response**308WithPostReverted
**response**300MultipleChoiceWithPost
**response**301MovedPermanentlyWithPost
**response**302MovedTemporarilyWithPost
**response**303SeeOtherWithPost
postRedirectToGetWithChunkedRequest
postRedirectToGetWithStreamedRequest

**URLReusedConnectionTest**
**postFailsWith**BufferedRequestForSmallRequest
**postFailsWith**BufferedRequestForLargeRequest
**postFailsWith**ChunkedRequestForSmallRequest
**postFailsWith**ChunkedRequestForLargeRequest
**postFailsWith**FixedLengthRequestForSmallRequest
**postFailsWith**FixedLengthRequestForLargeRequest

### URLZeroLengthConnectionTest
**zeroLength**Post
**zeroLength**Post_HTTP_2
**zeroLength**Put
**zeroLength**Put_HTTP_2

### CODE REVIEW

File: okhttp/src/test/java/okhttp3/URLConnectionUploadTest.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+import okhttp3.internal.RecordingOkAuthenticator;
+import okhttp3.internal.platform.Platform;
+import okhttp3.mockwebserver.MockResponse;
+import okhttp3.mockwebserver.MockWebServer;
+import okhttp3.mockwebserver.RecordedRequest;
+import okhttp3.mockwebserver.SocketPolicy;
+import okhttp3.testing.Flaky;
+import okhttp3.testing.PlatformRule;
+import okhttp3.tls.HandshakeCertificates;
+import okio.*;
```
Comment (yschimke) [inline okhttp/src/test/java/okhttp3/URLConnectionUploadTest.java:29]: Can you run `./gradlew spotlessApply` to cleanup

File: okhttp/src/test/java/okhttp3/URLConnectionUploadTest.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+  @Rule public final OkHttpClientTestRule clientTestRule = new OkHttpClientTestRule();
+  @Rule public final MockWebServer server = new MockWebServer();
+  private OkHttpClient client = clientTestRule.newClient();
+
+  @Before public void setUp() {
+    server.setProtocolNegotiationEnabled(false);
+  }
+
+  @Test public void chunkedUpload_byteByByte() throws Exception {
```
Comment (yschimke) [inline okhttp/src/test/java/okhttp3/URLConnectionUploadTest.java:76]: Is it worth parameterizing this test?

**Comment by yschimke** [comment]
No strong opinion from me, I'll defer to others on this.  But I'm curious what drove this change?

**Comment by swankjesse** [comment]
Each test is usually the intersection of 2 or more features. Organizing by feature may make it harder to find tests, not easier. Some tests here do upload and connection reuse, for example.

I'll second Yuri's question: what problem are you trying to solve?

File: okhttp/src/test/java/okhttp3/URLConnectionUploadTest.java (hunk 2 above)
Comment (klebertarcisio) [reply okhttp/src/test/java/okhttp3/URLConnectionUploadTest.java:76]: Yes, you're right. I believe that in this case it is not necessary. I just corrected it. I also removed unnecessary imports. Please tell me if you have any more suggestions. I will be happy to help.

**Comment by klebertarcisio** [comment]
Hi, thanks for the answers!

I used a tool to detect smells and this class was classified as **GodClass** because it has more than **500 lines of code** and has a lower **Tight Class Cohesion** than the average of the other classes in the system. 

_Do you think that 500 lines for OkHttp is a reasonable number?_ 

Given the motivations I mentioned, _do you think it is necessary to remove GodClass?_

I looked for similar tests to create new classes. _What would be a better strategy?_

File: okhttp/src/test/java/okhttp3/URLConnectionUploadTest.java (hunk 1 above)
Comment (klebertarcisio) [reply okhttp/src/test/java/okhttp3/URLConnectionUploadTest.java:21]: I removed unnecessary imports. Thanks.

**Comment by yschimke** [comment]
@klebertarcisio I don't think overall applying your personal metrics of code quality to a project is a high value use of your time.  

IMHO - This one concern about whether a test class is a GodClass is probably most relevant if you can see that almost all PRs involve changing these tests also.  I think that's the concern about the GodClass in a normal system, everyone ends up changing the same files with risky merges etc, because everything is muddled up there. But I think for metrics driven development, we would value this sort of refactor more for non-test classes.  

CallTest and URLConnectionTest are probably the most likely files to change for any given PR.  But mostly I think we just add new test cases there. On further thought my bigger concern is that a primary use of these tests is as a regression suite. We would lose this history between branches.

Look at `git diff okhttp_3.14.x:okhttp-tests/src/test/java/okhttp3/URLConnectionTest.java master:okhttp/src/test/java/okhttp3/URLConnectionTest.java` 

So I'm going to shut down this avenue, unless another maintainer steps in to actively support the change.

**Comment by klebertarcisio** [comment]
Hi @yschimke, I understood your point, thank you very much for your feedback.

**Pull request closed without merging**